File: lista/Python/questao11.py
"""
Crie uma classe abstrata Funcionario com um método abstrato
calcularSalario. Derive classes como FuncionarioHorista e FuncionarioAssalariado que
implementam calcularSalario de formas distintas.

"""
from abc import ABC, abstractmethod
class Funcionario(ABC):
    @abstractmethod
    def calcularSalario(self):
        pass
class FuncionarioHorista(Funcionario):
    def __init__(self, nome, horas_trabalhadas, valor_hora):
        self.nome = nome
        self.horas_trabalhadas = horas_trabalhadas
        self.valor_hora = valor_hora
    def calcularSalario(self):
        return self.horas_trabalhadas * self.valor_hora

class FuncionarioAssalariado(Funcionario):
    def __init__(self, nome, salario_mensal):
        self.nome = nome
        self.salario_mensal = salario_mensal

    def calcularSalario(self):
        return self.salario_mensal

def main():
    funcionario1 = FuncionarioHorista("Carlos", 100, 35)
    print(f"Salário de {funcionario1.nome}(Funcionario Horista): R${funcionario1.calcularSalario():.2f}")

    funcionario2 = FuncionarioAssalariado("Maria", 3500)
    print(f"Salário de {funcionario2.nome} (Funcionario Assalariado): R${funcionario2.calcularSalario():.2f}")

if __name__ == "__main__":
    main()

File: lista/GO/questao11/questao11.go
package main

import (
	"fmt"
)

type Funcionario interface {
	CalcularSalario() float64
}

type FuncionarioHorista struct {
	HorasTrabalhadas float64
	ValorHora        float64
}

func (f FuncionarioHorista) CalcularSalario() float64 {
	return f.HorasTrabalhadas * f.ValorHora
}

type FuncionarioAssalariado struct {
	SalarioMensal float64
}

func (f FuncionarioAssalariado) CalcularSalario() float64 {
	return f.SalarioMensal
}

func ExibirSalario(f Funcionario) {
	fmt.Printf("Salário: R$ %.2f\n", f.CalcularSalario())
}

func main() {
	funcionarioHorista := FuncionarioHorista{
		HorasTrabalhadas: 100,
		ValorHora:        35,
	}

	funcionarioAssalariado := FuncionarioAssalariado{
		SalarioMensal: 3500,
	}

	fmt.Println("Funcionário Horista:")
	ExibirSalario(funcionarioHorista)

	fmt.Println("Funcionário Assalariado:")
	ExibirSalario(funcionarioAssalariado)
}
